src/mono/mono/mini/mini-runtime.c

```diff
@@ -5395,3 +5401,33 @@ mini_alloc_jinfo (MonoJitMemoryManager *jit_mm, int size)
 		return (MonoJitInfo*)mono_mem_manager_alloc0 (jit_mm->mem_manager, size);
 	}
 }
+
+void
+mono_set_runtime_init_callback (MonoRuntimeInitCallback callback)
+{
+	runtime_init_callback = callback;
+}
+
+/*
+* Default implementation invoking runtime init callback in native-to-managed wrapper.
+*/
+void
+mono_invoke_runtime_init_callback (void)
+{
+	if (G_UNLIKELY (runtime_init_callback)) {
+		guint64 thread_id = mono_native_thread_os_id_get ();
+		if (runtime_init_callback && (runtime_init_thread_id == thread_id))
+			return;
+
+		while (mono_atomic_cas_i64 ((volatile gint64 *)&runtime_init_thread_id, (gint64)thread_id, (gint64)G_MAXUINT64) != (gint64)G_MAXUINT64)
+			g_usleep (1000);
+
+		if (runtime_init_callback) {
+			if (!mono_thread_info_current_unchecked ())
+				runtime_init_callback ();
+			runtime_init_callback = NULL;
+		}
+
+		mono_atomic_xchg_i64 ((volatile gint64 *)&runtime_init_thread_id, (gint64)G_MAXUINT64);
```

Comment:
What is supposed to happen here? When we get to this, won't `&runtime_init_thread_id` hold `False` (`(gint64)G_MAXUINT64) != (gint64)G_MAXUINT64`), why do we need need to set the value at `&runtime_init_thread_id` back to the max `G_MAXUINT64`?

Reply:
This is a lock free way to make sure only one threads gets to call the callback and initialize the runtime. So the runtime_init_thread_id holds the thread id of the current thread initializing the runtime. That is mainly needed to make sure you block recursive calls from the same thread, other threads are stuck on the compare and swap loop as long as the thread id is not G_MAXUINT64, when the thread initializing the runtime is done, it will set it back to G_MAXUINT64 that will unblock next thread, but since runtime_init_callback is NULL at that point it will jump past the call to callback and reset the runtime_init_thread_id to G_MAXUINT64 that will unblock the next thread etc. So G_MAXUINT64 indicates that the "lock" is free, using that value since we have seen 0 as a valid thread id on some platforms in the past.